src/pushsource/_impl/backend/koji_source.py

```diff
@@ -509,6 +521,94 @@ def _push_items_from_container_build(self, nvr, meta):
 
         return out
 
+    def _push_items_from_vmi_build(self, nvr, meta):
+        LOG.debug("Looking for virtual machine images on %s, %s", nvr, meta)
+
+        if not meta:
+            message = "Virtual machine image build not found in koji: %s" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        # VMI has the {'typeinfo': {'image':{}}} on extra
+        extra = meta.get("extra") or {}
+        typeinfo = extra.get("typeinfo") or {}
+        image = typeinfo.get("image")
+
+        if image is None:
+            message = "Build %s not recognized as a virtual machine image build" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        build_id = meta["id"]
+
+        archives = self._get_archives(build_id)
+
+        # Supported types of VMI archives
+        ami_types = ["raw", "raw-xz"]
+        azure_types = ["vhd", "vhd-compressed"]
+        cloud_types = {}
+
+        for k in ami_types:
+            cloud_types.setdefault(k, AmiPushItem)
+
+        for k in azure_types:
+            cloud_types.setdefault(k, VHDPushItem)
+
+        # Collect all image archives which are not containers (tar)
+        vmi_archives = [
+            elem
+            for elem in archives
+            if elem.get("btype") == "image" and elem.get("type_name") != "tar"
+        ]
+
+        out = []
+
+        # Generate the respective PushItem type for each archive
+        for archive in vmi_archives:
+            path = self._pathinfo.typedir(meta, archive["btype"])
+            item_src = os.path.join(path, archive["filename"])
+
+            extra = archive.get("extra") or {}
+            image = extra.get("image") or {}
+            arch = image.get("arch")
+
+            klass = cloud_types.get(archive["type_name"]) or VMIPushItem
+
+            rel_klass = AmiRelease if klass == AmiPushItem else VMIRelease
+
+            splitted_name = meta["name"].split("-")
+            if len(splitted_name) > 1:
+                product = "-".join(splitted_name[:-1])
+            else:
+                product = splitted_name[0]
+
+            release = rel_klass(
+                arch=arch,
+                date=meta.get("completion_time", "").split(" ")[0],
```

Comment:
Minor: if you want this field to default to an empty string, it's probably better to be more defensive like this:

```suggestion
                date=(meta.get("completion_time") or "").split(" ")[0],
```

Right now, with the way you've written it, if completion_time is present but null then it'll crash.

Reply:
Thanks!

Reply:
Fixed, PTAL again